pkg/updater/updater.go

```diff
@@ -96,52 +71,172 @@ func (u *updater) RunOnce(ctx context.Context) error {
 			continue
 		}
 
-		rackStatuses := cluster.Status.Racks
-		anyRackUpdated := false
 		for j := range rackRecs {
 			rackRec := &rackRecs[j]
-
-			if rackRec.Members == nil {
-				u.logger.Debug(ctx, "no members recommendation for rack", "rack", rackRec.Name)
-				continue
-			}
-
 			rack := findRack(rackRec.Name, cluster.Spec.Datacenter.Racks)
 			if rack == nil {
 				u.logger.Debug(ctx, "rack specified in recommendation was not found in cluster's data center",
 					"rack", rackRec.Name, "cluster", cluster.Name, "data center", cluster.Spec.Datacenter.Name)
 				continue
 			}
 
-			rackStatus := rackStatuses[rack.Name]
-			requestedMembers := rackStatus.Members
-			readyMembers := rackStatus.ReadyMembers
+			applyRackRec(rack, rackRec)
+		}
 
-			if requestedMembers != readyMembers {
-				u.logger.Debug(ctx, "not applying recommendation: rack's requested members aren't equal to ready",
-					"rack", rack.Name, "requested members", requestedMembers, "ready members", readyMembers)
-				continue
-			} else if rack.Members == *rackRec.Members {
-				u.logger.Debug(ctx, "not applying recommendation: rack's spec members are equal to target",
-					"rack", rack.Name, "requested members", requestedMembers,
-					"target members", *rackRec.Members)
-				continue
-			}
+		if err = u.updateScyllaCluster(ctx, cluster, sca.Status.Recommendations); err != nil {
+			return err
+		}
+		if err = u.updateSCAStatus(ctx, sca); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
 
-			rack.Members = *rackRec.Members
-			anyRackUpdated = true
-			u.logger.Info(ctx, "members recommendation for rack applied",
-				"rack", rackRec.Name, "members", rack.Members)
+func filterSCAs(scas *v1alpha1.ScyllaClusterAutoscalerList) []v1alpha1.ScyllaClusterAutoscaler {
+	filteredSCAs := make([]v1alpha1.ScyllaClusterAutoscaler, 0)
+	for _, sca := range scas.Items {
+		if sca.Spec.UpdatePolicy != nil && sca.Spec.UpdatePolicy.UpdateMode == v1alpha1.UpdateModeAuto &&
+			sca.Status.UpdateStatus != nil && *sca.Status.UpdateStatus == v1alpha1.UpdateStatusOk {
+			filteredSCAs = append(filteredSCAs, sca)
 		}
+	}
 
-		if anyRackUpdated {
-			if err = u.client.Update(ctx, cluster); err != nil {
-				return err
+	return filteredSCAs
+}
+
+func (u *updater) recommendationExpired(ctx context.Context, sca *v1alpha1.ScyllaClusterAutoscaler) bool {
+	if recExpTime := sca.Spec.UpdatePolicy.RecommendationExpirationTime; !sca.Status.LastUpdated.IsZero() &&
+		recExpTime != nil && time.Now().Sub(sca.Status.LastUpdated.Time) > recExpTime.Duration {
+		u.logger.Info(ctx, "sca's recommendation expired", "sca", sca.Name, "namespace", sca.Namespace)
```

Comment:
I would still move logging outside of these predicates and instead writing `if (true) return true else return false` i would simply return if condition.